query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/QueryRequestToPinotSQLConverter.java

```diff
@@ -335,39 +433,53 @@ private String getLogicalColumnName(Expression expression) {
     }
   }
 
-  private LiteralConstant[] convertExpressionToMapLiterals(Expression expression) {
+  private LiteralConstant[] convertExpressionToMapLiterals(Expression rhs, Expression lhs) {
     LiteralConstant[] literals = new LiteralConstant[2];
-    if (expression.getValueCase() == LITERAL) {
-      LiteralConstant value = expression.getLiteral();
+    String[] literalArguments = new String[] {"", ""};
+
+    if (rhs.getValueCase() == LITERAL) {
+      LiteralConstant value = rhs.getLiteral();
+      // backward compatibility
       if (value.getValue().getValueType() == ValueType.STRING_ARRAY) {
-        for (int i = 0; i < 2 && i < value.getValue().getStringArrayCount(); i++) {
-          literals[i] =
-              LiteralConstant.newBuilder()
-                  .setValue(
-                      Value.newBuilder().setString(value.getValue().getStringArray(i)).build())
-                  .build();
+        literalArguments[0] = value.getValue().getStringArray(0);
+        if (value.getValue().getStringArrayCount() > 1) {
+          literalArguments[1] = value.getValue().getStringArray(1);
         }
+      } else if (value.getValue().getValueType() == ValueType.STRING) {
```

Comment:
It is used for attribute expression when a map attribute is present in the filter. Not necessarily contains kep operator.

Reply:
The list is only marginally better! The suggestion was actually asking to use a purpose-built data structure.
e.g.
```java
@Value
@Builder
class MapLiteralPair {
  LiteralConstant key;
  LiteralConstant value;
}
```


BUT - I'm still of the opinion this function doesn't need to change. If it's an attribute expression, we have the subpath - it's far less complex to just read it in directly as shown in my previous example

Reply:
>BUT - I'm still of the opinion this function doesn't need to change. If it's an attribute expression, we have the subpath - it's far less complex to just read it in directly as shown in my previous example

ValueType is of different format over here.

Reply:
> ValueType is of different format over here.

I'm not sure what you mean - assuming we revert the filter changes (because they've been replaced by preprocessing), CONTAINS_KEY and CONTAINS_KEY_VALUE are the only places this function is called and they would still be parsed in the same way as they had been.

Reply:
Checking again after pre processing.

Reply:
Reverted the changes.

Reply:
resolved